tests/bluetooth/bsim_bt/bsim_test_l2cap/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,563 @@
+	int ret = bt_l2cap_chan_send(chan, buf);
+
+	LOG_DBG("bt_l2cap_chan_send returned: %i", ret);
+
+	if (ret < 0) {
+		LOG_DBG("Error: send failed error: %i", ret);
+		channels[chan_idx].buf = 0;
+		net_buf_unref(buf);
+	}
+}
+
+static void test_peripheral_main(void)
+{
+	int err;
+
+	LOG_DBG("*L2CAP ECRED Peripheral started*");
+	init_workqs();
+	err = bt_enable(NULL);
+	if (err) {
+		FAIL("Can't enable Bluetooth (err %d)", err);
+		return;
+	}
+	LOG_DBG("Peripheral Bluetooth initialized.");
+
+	LOG_DBG("Connectable advertising...");
+	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
+	if (err) {
+		FAIL("Advertising failed to start (err %d)", err);
+		return;
+	}
+	LOG_DBG("Advertising started.");
+
+	LOG_DBG("Peripheral waiting for connection...");
+	WAIT_FOR_FLAG_SET(is_connected);
+	LOG_DBG("Peripheral Connected.");
+
+	/* Wait a bit to ensure that all LLCP have time to finish */
+	k_sleep(K_MSEC(1000));
+
+	register_l2cap_server();
+
+	connect_num_channels(L2CAP_CHANNELS);
+
+	k_sleep(K_MSEC(500));
+	for (int i = 0; i < SDU_SEND_COUNT; i++) {
+		channels[0].itteration = i;
+		channels[0].bytes_to_send = DATA_MTU - 500;
+		k_work_submit_to_queue(&my_work_q0, &channels[0].work);
+		channels[1].itteration = i;
+		channels[1].bytes_to_send = DATA_MPS - 2;
+		k_work_submit_to_queue(&my_work_q1, &channels[1].work);
+		k_sleep(K_MSEC(5000));
+	}
+	disconnect_all_channels();
+	WAIT_FOR_FLAG_UNSET(is_connected);
+	PASS("L2CAP ECRED Peripheral tests Passed");
+	bs_trace_silent_exit(0);
+}
+
+static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
+			 struct net_buf_simple *ad)
+{
+	struct bt_le_conn_param *param;
+	int err;
+
+	err = bt_le_scan_stop();
+	if (err) {
+		FAIL("Stop LE scan failed (err %d)", err);
+		return;
+	}
+
+	param = BT_LE_CONN_PARAM_DEFAULT;
+	err = bt_conn_le_create(addr, BT_CONN_LE_CREATE_CONN, param, &default_conn);
+	if (err) {
+		FAIL("Create conn failed (err %d)", err);
+		return;
+	}
+}
+
+static void test_central_main(void)
+{
+	struct bt_le_scan_param scan_param = {
+		.type = BT_LE_SCAN_TYPE_ACTIVE,
+		.options = BT_LE_SCAN_OPT_NONE,
+		.interval = BT_GAP_SCAN_FAST_INTERVAL,
+		.window = BT_GAP_SCAN_FAST_WINDOW,
+	};
+
+	int err;
+
+	LOG_DBG("*L2CAP ECRED Central started*");
+
+	err = bt_enable(NULL);
+	if (err) {
+		FAIL("Can't enable Bluetooth (err %d)\n", err);
+		return;
+	}
+	LOG_DBG("Central Bluetooth initialized.\n");
+
+	err = bt_le_scan_start(&scan_param, device_found);
+	if (err) {
+		FAIL("Scanning failed to start (err %d)\n", err);
+		return;
+	}
+
+	LOG_DBG("Scanning successfully started\n");
+
+	LOG_DBG("Central waiting for connection...\n");
+	WAIT_FOR_FLAG_SET(is_connected);
+	LOG_DBG("Central Connected.\n");
+	register_l2cap_server();
+	while(num_connect_chans < L2CAP_CHANNELS)
+	{
+		(void)k_sleep(K_MSEC(1));
+	}
+	LOG_DBG("Both l2cap channels connected, test started\n");
+	while(num_connect_chans > 0)
+	{
+		(void)k_sleep(K_MSEC(1));
+	}
```

Comment:
Consider using `K_SEM_DEFINE`, `k_sem_give` and `k_sem_take` to wait for X number of connections and disconnections.

Reply:
Correct. 

It will basically give the same result, but less busy-polling (either is fine since this is just a BSIM test).